src/sync/partial_sync.cpp

```diff
@@ -56,6 +83,153 @@ void update_schema(Group& group, Property matches_property)
+                                    std::string const& query,
+                                    std::string const& matches_property,
+                                    ObjectSchema const& result_sets_schema)
+{
+    TableRef table = ObjectStore::table_for_object_type(realm->read_group(), result_sets_type_name);
+    size_t name_idx = result_sets_schema.property_for_name("name")->table_column;
+    auto existing_row_ndx = table->find_first_string(name_idx, name);
+    if (existing_row_ndx == npos)
+        return false;
+
+    Object existing_object(realm, result_sets_schema, table->get(existing_row_ndx));
+
+    CppContext context;
+    std::string existing_query = any_cast<std::string>(existing_object.get_property_value<util::Any>(context, "query"));
+    if (existing_query != query)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different query.");
+
+    std::string existing_matches_property = any_cast<std::string>(existing_object.get_property_value<util::Any>(context, "matches_property"));
+    if (existing_matches_property != matches_property)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different result type.");
+
+    return true;
+}
+
+class WorkQueue {
+public:
+    ~WorkQueue()
+    {
+        reset();
+    }
+
+    void reset()
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = true;
+        }
+        m_cv.notify_one();
+
+        if (m_thread.joinable())
+            m_thread.join();
+
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = false;
+        }
+    }
+
+    void enqueue(std::function<void()> function)
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_queue.push_back(std::move(function));
+
+            if (m_stopped)
+                create_thread();
+        }
+        m_cv.notify_one();
+    }
+
+private:
+    void create_thread()
+    {
+        if (m_thread.joinable())
+            m_thread.join();
+
+        m_thread = std::thread([this] {
+            std::vector<std::function<void()>> queue;
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            while (!m_stopping &&
+                   m_cv.wait_for(lock, std::chrono::milliseconds(500),
+                                 [&] { return !m_queue.empty() || m_stopping; })) {
+
+                swap(queue, m_queue);
+
+                lock.unlock();
+                for (auto& f : queue)
+                    f();
+                queue.clear();
+                lock.lock();
+            }
+
+            m_stopped = true;
+        });
+        m_stopped = false;
+    }
+
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<std::function<void()>> m_queue;
+    std::thread m_thread;
+    bool m_stopping = false;
+    bool m_stopped = true;
+};
+
+static auto& work_queue = *new WorkQueue;
```

Comment:
A single global work queue seems problematic. It means that a write transaction on one Realm can block creating partial sync subscriptions for different Realms. I can't think of a case where this would outright deadlock, but it feels like the sort of thing that'll inevitably cause problems.

Reply:
The fact that you can't know when the Realm will actually be closed after registering a partial sync query also seems like a potential problem, but I don't know what to do about that.

Reply:
> A single global work queue seems problematic. It means that a write transaction on one Realm can block creating partial sync subscriptions for different Realms.

The only problem that came to mind was the potential for unexpected delays when subscribing to multiple Realms, which doesn't see great but also seems unlikely to be problematic. We could avoid that by having the queue be per-`RealmCoordinator` instead if it makes you nervous.

> The fact that you can't know when the Realm will actually be closed after registering a partial sync query also seems like a potential problem, but I don't know what to do about that.

I'm not sure what you mean here. The background Realm will be closed at the point the `Subscription` moves to either the error or complete state.

Reply:
Hm... I'm not sure I fully understand this logic. The way I read the code, the background Realm will be closed as soon as the object is written, regardless of what happens to the subscription object. Moving the subscription to the complete/initialized/whatever state happens on the original thread, does it no?

Reply:
That's correct. The background Realm is only used for validating and creating the subscription. It is then closed.

Reply:
I think this might cause problems for Java. Especially in test setups where we delete Realms after they are closed. We currently use `SyncSessionStopPolicy::Immediately` for the same reason, but this queue seems to outlive the lifecycle of the Realm in nondeterministic ways which look like it could crashes if the underlying Realm is deleted in the meantime.

Reply:
After looking at `partial_sync::reset_for_testing()`, that seems make sure the above doesn't happen, correct?

Reply:
I moved the work queue to be owned by the `RealmCoordinator`, so it will be torn down as soon as the last reference to a given Realm is dropped.

And yes `partial_sync::reset_for_testing()` existed to prevent that problem with the earlier singleton queue. I removed it since it shouldn't be necessary with the new ownership model.